hal/src/rtl872x/ble_hal.cpp

```diff
@@ -2104,22 +2166,42 @@ bool BleGap::isPaired(hal_ble_conn_handle_t connHandle) {
 }
 
 int BleGap::waitState(BleGapDevState state, system_tick_t timeout, bool forcePoll) {
+    LOCAL_DEBUG("Enter waitState");
+    bool ret = false;
+    RtlGapDevState s;
+    if (le_get_gap_param(GAP_PARAM_DEV_STATE, &s.state) == GAP_CAUSE_SUCCESS) {
+        LOCAL_DEBUG("waitState(), get: %02X, now: %02X", s.raw, state_.raw);
+        if (!(state_.raw == s.raw)) {
+            state_.raw = s.raw;
+        }
+        if (BleGapDevState(getState()).copySetParamsFrom(state) == state) {
+            ret = true;
+        }
+    }
+    if (ret) {
+        LOCAL_DEBUG("Target state is already set, return.");
+        return SYSTEM_ERROR_NONE;
+    }
+
+    auto now = hal_timer_millis(nullptr);
+    auto end = now + timeout;
     auto current = BleGapDevState(getState());
     if (forcePoll || current.isNotInitialized() || !btStackStarted_) {
+        LOCAL_DEBUG("force poll, t: %d", end - now);
         // Poll, as some of the initial events are missing
         RtlGapDevState s;
-        auto start = hal_timer_millis(nullptr);
-        while (hal_timer_millis(nullptr) - start < timeout) {
+        while (now < end) {
             CHECK_RTL(le_get_gap_param(GAP_PARAM_DEV_STATE, &s.state));
             if (BleGapDevState(s).copySetParamsFrom(state) == state) {
                 return SYSTEM_ERROR_NONE;
             }
             HAL_Delay_Milliseconds(BLE_WAIT_STATE_POLL_PERIOD_MS);
+            now = hal_timer_millis(nullptr);
         }
     } else {
-        auto end = hal_timer_millis(nullptr) + timeout;
-        for (auto now = hal_timer_millis(nullptr); now < end; now = hal_timer_millis(nullptr)) {
+        for (now = now; now < end; now = hal_timer_millis(nullptr)) {
```

Comment:
What is the intent behind `now = now`?

Reply:
This is just to make the code more readable. I did observe that due to thread scheduling if `now` is set in `for()`, the time elapsed between it and when `end` is set will make `now < end` false. Note that in BLE scan method we have `if (waitState(BleGapDevState().scan(GAP_SCAN_STATE_START), 10, true) && waitState(BleGapDevState().scan(GAP_SCAN_STATE_SCANNING), 10, true)) {`, which sets the timeout to 10ms. It's easily to make `now < end` false so that we may lose a chance to check the state.